src/wallet/wallet.h

```diff
@@ -477,16 +475,46 @@ class CWalletTx
         fInMempool = false;
         nChangeCached = 0;
         nOrderPos = -1;
+        tx_state.hashBlock = uint256();
+        tx_state.nIndex = 0;
+        setUnconfirmed();
     }
 
     CTransactionRef tx;
-    uint256 hashBlock;
-    /* An nIndex == -1 means that hashBlock (in nonzero) refers to the earliest
-     * block in the chain we know this or any in-wallet dependency conflicts
-     * with. Older clients interpret nIndex == -1 as unconfirmed for backward
-     * compatibility.
+
+    /* New transactions start in the UNCONFIRMED. At BlockConnected,
+     * they will transition to CONFIRMED. In case of reorg, at BlockDisconnected,
+     * they rollback to UNCONFIRMED. If we detect a conflicting transaction at
+     * block connection, we update conflited tx and its dependencies as CONFLICTED.
+     * In tx isn't confirmed and outside of mempool, the user may switch it to ABANDONED
+     * thanks to the abandontransaction call. This last state may be override by a CONFLICTED
+     * or CONFIRMED transition.
      */
-    int nIndex;
+    enum TxState {
+        UNCONFIRMED,
+        CONFIRMED,
+        CONFLICTED,
+        ABANDONED
+    };
+
+    /* TxConfirmation includes tx state and a pair of block hash/block index at which tx has been confirmed.
+     * This pair is both 0 if tx hasn't confirmed yet. Meaning of these fields changes with CONFLICTED state
+     * where they instead point to block hash and index of the deepest conflicting tx.
+     */
+    struct TxConfirmation {
+        uint256 hashBlock;
+        /* At serialization/deserialization, an nIndex == -1 means that hashBlock refers to
+         * the earliest block in the chain we know this or any in-wallet dependency conflicts
+         * with, but if hashBlock is null it means transaction is abandoned. An nIndex >= 0
```

Comment:
Currently, `AbandonTransaction` set `nIndex` as -1 and this value is kept at serialization/deserialization. I'm not sure if it make sense for older client, because an abandoned transaction can be also confirmed/unconfirmed/conflicted at same time. Nevertheless, I've tried to keep same behavior for now, so `hashBlock` set as ABANDON_HASH is used to avoid deserialization ambiguity with conflicted.

I've clarified this comment, the ABANDON_HASH one and also reset abandon txn `hashBlock` at deserialization. ABANDON_HASH should be scoped at serialization/deserialization.